Mapping over a list and returning within an `If` branch

<!-- You can use `v doctor` to fill up the next fields -->
**V version: V 0.2.2 977a625, timestamp: 2021-02-23 10:05:59 +0200** 
**OS: macos, Mac OS X, 10.14.6, 18G8022** 

**What did you do?**
I executed:
```
fn main() {
	num_string := '2 3'

	assigned := if num_string.len > 1 {
		num_string.split(' ').map(it.int())
	} else {
		[789]
	}

	println(assigned)
}
```

It seems to be an issue of mapping over the list and returning the result inside an `if` branch.

It works fine without the mapping to ints:
```
fn main() {
	num_string := '2 3'
	assigned := if num_string.len > 1 {
		num_string.split(' ')
	} else {
		['789']
	}

	println(assigned)
}
// ['2', '3']
```

and with the mapping to ints but this time not within an `if` branch:
```
fn main() {
	num_string := '2 3'
	assigned := num_string.split(' ').map(it.int())
	println(assigned)
}
// [2, 3]
```

**What did you expect to see?**
[2, 3]
 
**What did you see instead?**
```
==================

/var/folders/wd/3x4r7ylx7vj7mc9pyl58f__00000gn/T/v/CrashExample.14390554357206708833.tmp.c:8623:108: error: expected expression
        Array_int assigned = (num_string.len > 1 ? (    Array_string _t69_orig = string_split(num_string, _SLIT(" "));
                                                                                                                     ^
/var/folders/wd/3x4r7ylx7vj7mc9pyl58f__00000gn/T/v/CrashExample.14390554357206708833.tmp.c:8623:108: error: expected ':'
        Array_int assigned = (num_string.len > 1 ? (    Array_string _t69_orig = string_split(num_string, _SLIT(" "));
                                                                                                                     ^
                                                                                                                     : 
/var/folders/wd/3x4r7ylx7vj7mc9pyl58f__00000gn/T/v/CrashExample.14390554357206708833.tmp.c:8623:43: note: to match this '?'
        Array_int assigned = (num_string.len > 1 ? (    Array_string _t69_orig = string_split(num_string, _SLIT(" "));
                                                 ^
/var/folders/wd/3x4r7ylx7vj7mc9pyl58f__00000gn/T/v/CrashExample.14390554357206708833.tmp.c:8623:108: error: expected expression
...
==================
(Use `v -cg` to print the entire error message)

builder error: 
==================
C error. This should never happen.

If you were not working with C interop, please raise an issue on GitHub:
```

--- a/vlib/v/gen/c/cgen.v
+++ b/vlib/v/gen/c/cgen.v
@@ -4069,6 +4069,27 @@ fn (mut g Gen) concat_expr(node ast.ConcatExpr) {
 	}
 }
 
+fn (mut g Gen) need_tmp_var_in_if(node ast.IfExpr) bool {
+	if node.is_expr && node.typ != table.void_type && node.typ != 0 && g.inside_ternary == 0 {
+		for branch in node.branches {
+			if branch.stmts.len == 1 {
+				if branch.stmts[0] is ast.ExprStmt {
+					stmt := branch.stmts[0] as ast.ExprStmt
+					if stmt.expr is ast.CallExpr {
+						if stmt.expr.is_method {
+							left_sym := g.table.get_type_symbol(stmt.expr.receiver_type)
+							if left_sym.kind in [.array, .array_fixed, .map] {
+								return true
+							}
+						}
+					}
+				}
+			}
+		}
+	}
+	return false
+}
+
 fn (mut g Gen) if_expr(node ast.IfExpr) {
 	if node.is_comptime {
 		g.comp_if(node)
@@ -4080,15 +4101,7 @@ fn (mut g Gen) if_expr(node ast.IfExpr) {
 	// easier to use a temp var, than do C tricks with commas, introduce special vars etc
 	// (as it used to be done).
 	// Always use this in -autofree, since ?: can have tmp expressions that have to be freed.
-	first_branch := node.branches[0]
-	needs_tmp_var := node.is_expr && (g.is_autofree || (g.pref.experimental
-		&& (first_branch.stmts.len > 1 || (first_branch.stmts[0] is ast.ExprStmt
-		&& (first_branch.stmts[0] as ast.ExprStmt).expr is ast.IfExpr))))
-	/*
-	needs_tmp_var := node.is_expr &&
-		(g.autofree || g.pref.experimental) &&
-		(node.branches[0].stmts.len > 1 || node.branches[0].stmts[0] is ast.IfExpr)
-	*/
+	needs_tmp_var := g.need_tmp_var_in_if(node)
 	tmp := if needs_tmp_var { g.new_tmp_var() } else { '' }
 	mut cur_line := ''
 	if needs_tmp_var {

--- a/vlib/v/gen/c/cgen.v
+++ b/vlib/v/gen/c/cgen.v
@@ -4070,7 +4070,10 @@ fn (mut g Gen) concat_expr(node ast.ConcatExpr) {
 }
 
 fn (mut g Gen) need_tmp_var_in_if(node ast.IfExpr) bool {
-	if node.is_expr && node.typ != table.void_type && node.typ != 0 && g.inside_ternary == 0 {
+	if node.is_expr && g.inside_ternary == 0 {
+		if g.is_autofree {
+			return true
+		}
 		for branch in node.branches {
 			if branch.stmts.len == 1 {
 				if branch.stmts[0] is ast.ExprStmt {

--- a/vlib/v/tests/if_expression_test.v
+++ b/vlib/v/tests/if_expression_test.v
@@ -167,3 +167,16 @@ fn test_multi_if_expr_with_infix() {
 	a := if 1 == 0 { 1 } else if 1 == 0 { 2 } else { 3 } + 4
 	assert a == 7
 }
+
+fn test_if_expr_with_array_map() {
+	num_string := '2 3'
+
+	assigned := if num_string.len > 1 {
+		num_string.split(' ').map(it.int())
+	} else {
+		[789]
+	}
+
+	println(assigned)
+	assert assigned == [2, 3]
+}

